javaslang/src/main/java/javaslang/collection/TreeMap.java

```diff
@@ -477,6 +505,98 @@ public boolean isEmpty() {
         return TreeSet.ofAll(keyComparator(), iterator().map(Tuple2::_1));
     }
 
+    @Override
+    public Iterator<Tuple2<K, V>> descendingIterator() {
+        return entries.descendingIterator();
+    }
+
+    @Override
+    public SortedSet<K> descendingKeySet() {
+        return descendingMap().keySet();
+    }
+
+    @Override
+    public SortedMap<K, V> descendingMap() {
+        return new TreeMap<>(entries.descendingTree());
+    }
+
+    @Override
+    public SortedMap<K, V> subMap(Option<K> fromKey, boolean fromInclusive, Option<K> toKey, boolean toInclusive) {
+        return new TreeMap<>(
+                entries.subTree(
+                    fromKey.map(k -> Tuple.of(k, null)), fromInclusive,
```

Comment:
Not sure I follow, what do you mean by "half Map"?. 

We need to store both the key and the value, but we only search by key.

Both `TreeMap` and `TreeSet` use `RedBlackTree` as the backing structure.  `RedBlackTree` only understands about node values and a comparators. `TreeMap` stores the key/value pairs as `Tuple2` objects, and modifies the key comparator to work on those `Tuple2` objects. That updated comparator only uses the key part of the tuple. The value part is unknown.